### PR TITLE
Generate error on ambiguous generic types

### LINKED ISSUE #152: Type inference should have appropriate errors for ambiguous types
In circumstances where generics don't have enough information to resolve to a concrete type, type inference should have an error about ambiguous types. For example:
```qsharp
let x = Length([]);
```
The type of the array is not known, so the generic on `Length` cannot be made concrete. This does not interfere with the evaluator, which does not care about generics, but could become a problem for code generation.

### PR DESCRIPTION
**Pull request by sezna**
Closes #152 

The compiler previously did not track span information for types in the
type engine, and also lacked a mechanism to disambiguate types originating
from divergent expressions from regularly generated types. In order to
generate this error message, I had to add a `ty_meta` mapping that tracks
both span information, in the case of a normal, non-divergent type, and its
divergence.

Spans are only needed for non-divergent expressions because divergent exprs
don't generate ambiguous type warnings: their types are _intentionally_
unresolved.

### CODE REVIEW

**Review by kuzminrobin — APPROVED** [review]

File: compiler/qsc_frontend/src/typeck/infer.rs
Comment (samarsha) [inline compiler/qsc_frontend/src/typeck/infer.rs:180]: I would be more specific and call this something like `TySource`.

File: compiler/qsc_frontend/src/typeck.rs
Comment (samarsha) [inline compiler/qsc_frontend/src/typeck.rs:107]: Maybe just `Ambiguous`? (Or at least `AmbiguousTy` to be consistent with the other error names.)

File: compiler/qsc_frontend/src/typeck/infer.rs
Comment (swernli) [inline compiler/qsc_frontend/src/typeck/infer.rs:399]: I can't comment on the exact line as it's too far outside the diff, but the comment pointing to issue 152 above here can be removed with this fix.

**Review by kuzminrobin — APPROVED** [review]
Approving, but my approval is UNlikely to unblock this PR since I'm not a code owner.

File: npm/test/basics.js
Comment (swernli) [inline npm/test/basics.js:79]: This may have already been discussed, but it looks like `checkCode` is deprecated and the messages says we should use the langauge service instead. Is that something we can do yet, or will we need to use this until we have langauge service integration farther along?

**Review by swernli — APPROVED** [review]

**Comment by idavis** [comment]
@billti @cesarzc - this is blocked on npm and wasm approval which you both have IIRC

**Review by billti — APPROVED** [review]

**Pull request merged**
